---
title: "R_Assignment"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##import fang and snp files using the right formate for different types of variables
```{r}
library(tidyverse)
fang <- read_tsv("./fang_et_al_genotypes.txt",col_names=FALSE)
snp <- read_tsv("./snp_position.txt",col_names=TRUE,cols(SNP_ID = col_character(),
  cdv_marker_id = col_double(),
  Chromosome = col_character(),
  Position = col_number(),
  alt_pos = col_character(),
  mult_positions = col_character(),
  amplicon = col_character(),
  cdv_map_feature.name = col_character(),
  gene = col_character(),
  `candidate/random` = col_character(),
  Genaissance_daa_id = col_double(),
  Sequenom_daa_id = col_double(),
  count_amplicons = col_double(),
  count_cmf = col_double(),
  count_gene = col_double()))
```

##inspection of files
```{r}
file.info("fang_et_al_genotypes.txt")
file.info("snp_position.txt")
str(fang)
str(snp)
is.data.frame(fang)
is.data.frame(snp)
nrow(fang)
nrow(snp)
ncol(fang)
ncol(snp)
view(fang)
view(snp)
```
##select out useful columns
```{r}
fang_sub <- fang[ 0:2783 , 3:986]
snp_sub <- select (snp, SNP_ID, Chromosome, Position)
maize <- filter(fang_sub,`X3` %in% c("Group","ZMMIL", "ZMMMR" ,"ZMMLR"))
teosinte <- filter(fang_sub, `X3` %in% c("Group","ZMPBA","ZMPIL","ZMPJA"))
maize <- maize[ , 2:984]
teosinte <- teosinte[ ,2:984]
```
##transpose data, as data.frame instead of matrix, specify stringsasfactors=false, strings will maintain as strings
```{r}
maize_transposed <- as.data.frame(t(maize),stringsAsFactors = FALSE)
teosinte_transposed <- as.data.frame(t(teosinte),stringsAsFactors = FALSE)
names(maize_transposed)[1]<-"SNP_ID"
names(teosinte_transposed)[1]<-"SNP_ID"
```
##join the two files together based on the common columns
```{r}
maize_joined <- merge(snp_sub,maize_transposed)
teosinte_joined <- merge(snp_sub,teosinte_transposed)

```
##sort by increasing chromosome position values for maize dataset 
```{r}
maize_increasing <- arrange(maize_joined,Position)
for(i in 1:10){
  temp <- filter(maize_increasing, Chromosome == i)
  write.table(temp, paste("./maize/maize_chr",i,"_incr.txt",sep="\t"),col.names = FALSE, row.names = FALSE, quote = FALSE)
}

teosinte_increasing <- arrange(teosinte_joined, Position)
for(i in 1:10){
  temp <- filter(teosinte_increasing, Chromosome == i)
  write.table(temp, paste("./teosinte/teosinte_chr",i,"_incr.txt",sep="\t"),col.names = FALSE, row.names = FALSE, quote = FALSE)
}

```
##replace missing values by -/-
```{r}

maize_alt <- as.matrix(maize_joined)
maize_alt <- lapply(maize_joined,gsub,pattern = "?", replacement = "-", fixed = TRUE)
maize_alt <- as_tibble(maize_alt)
maize_alt$Position <- as.numeric(maize_alt$Position)

teosinte_alt <- as.matrix(teosinte_joined)
teosinte_alt <- lapply(teosinte_joined, gsub, pattern = "?", replacement = "-", fixed = TRUE)
teosinte_alt <- as_tibble(teosinte_alt)
teosinte_alt$Position<- as.numeric(teosinte_alt$Position)

```
##sort by decreasing value 
```{r}
teosinte_decreasing <- arrange(teosinte_alt,-Position)
maize_decreasing <- arrange(maize_alt, -Position)
```
##sort by choromosome value for teosinte/maize dataset
```{r}

for(i in 1:10){
  temp <- filter(maize_decreasing, Chromosome == i)
  write.table(temp, paste("./maize/maize_chr",i,"_decr.txt",sep="\t"),col.names = FALSE, row.names = FALSE, quote = FALSE)
}


for(i in 1:10){
  temp <- filter(teosinte_decreasing, Chromosome == i)
  write.table(temp, paste("./teosinte/teosinte_chr",i,"_decr.txt",sep="\t"),col.names = FALSE, row.names = FALSE, quote = FALSE)
}



```

##snps per chromosome & distribution for each chromosome
```{r}
fang_sub1 <- fang[0:2783 , 4:986]
fang_trans1 <- as.data.frame(t(fang_sub1),stringsAsFactors=FALSE)
names(fang_trans1)[1]<-"SNP_ID"
fang_merged1 <- merge(snp_sub,fang_trans1)
ggplot(fang_merged1,aes(x=as.character(Chromosome))) + geom_histogram(bins=12,stat = "count", aes(fill = Chromosome))

ggplot(fang_merged1, aes(x=as.numeric(Position))) +
  geom_density(aes(fill = as.character(Chromosome))) + facet_wrap(~
  as.character(Chromosome)) + theme_bw()
```
##creating graphs based on homozygous/heterozygous(melt first)
```{r}
library(reshape2)
fang_withHeader <- read_tsv("./fang_et_al_genotypes.txt")
fang_melted <- melt(fang_withHeader,id.var ="Sample_ID",measure.vars=names(fang_withHeader[4:986]))
new <- fang_melted %>% as.matrix()
new [new %in% c("A/A","T/T","C/C","G/G")] <- "homozygous"
new [new %in% c("A/G","A/T","C/G","C/T","G/T","A/C")] <- "heterozygous"
new [new %in% c("?/?")] <-"missing"
new <- as.data.frame(new, stringsAsFactors = FALSE)
new %>% ggplot() + geom_bar(aes(x = Sample_ID, fill=value), position = "fill")
```

##byGroup
```{r}
fang_melted2<- melt(fang_withHeader,id.var = "Group",measure.vars = names(fang_withHeader[4:986]))
new2 <- fang_melted2 %>% as.matrix()
new2 [new2 %in% c("A/A","T/T","C/C","G/G")] <- "homozygous"
new2 [new2 %in% c("A/G","A/T","C/G","C/T","G/T","A/C")] <- "heterozygous"
new2 [new2 %in% c("?/?")] <-"missing"
new2 <- as.data.frame(new2,stringsAsFactors =FALSE)
new2 %>% ggplot() + geom_bar(aes(x = Group, fill=value), position = "fill")
```

##visualize another feature of the dataset: Visualize the distribution of the snps on each different chromosomes. 
```{r}
ggplot(data=snp) + geom_point(mapping=aes(x=Chromosome,y=Position), color="blue")
```
